### PR TITLE
Windows fixes

### PR DESCRIPTION
**Pull request by Cyan4973**
This started as a small patch for #1817 : 
the function protecting users from the scenario where both source and dictionary are the same file would always be triggered within the Windows binary.
This is because file identification by inode does not seem to work on Windows, so it's better to just compare file names, which is an existing code path for Visual, just extend it to `mingw`.

But then the questions was : why has this problem not already been detected by tests ? Well, because there is no runtime test on Appveyor. So let's just enable them.

And then, it became a rabbit hole. Many little things (corner cases mostly, new features too) would start failing, making it impossible to complete the runtime on Windows. Fixing them one by one resulted into this more generous patch.

- Runtime tests would not start on MinGW environment
- Tests would fail console detection inside `mintty` environment
- `nul` and `NUL` trigger different behaviors on Windows
- Test mode `-t` was not working on Windows
   + optimized test mode, so that it doesn't open output file nor `fwrite()` any data
- Tests for `--output-dir-flat` were failing on Windows
- `harness` tests for the educational decoder would not compile
- `zstd file -D dict` was always failing on Windows
- Improved speed of `make check`
- Scores of cast warnings on Windows + MinGW (failing `-Werror`)
- Enable `make check` on Appveyor

### CODE REVIEW

**Comment by Cyan4973** [comment]
There is still a pb to fix : 
in `util.c`, a function pointer is casted into a different function pointer.
This is fine for Visual, but not for `gcc9`, which complains that these functions have different prototypes.
Using `void*` as an intermediate is fine for `gcc9` (and `clang`). But it's no longer okay for Visual...

File: programs/fileio.c [hunk 1]
```diff
@@ -671,56 +686,50 @@ int FIO_checkFilenameCollisions(const char** filenameTable, unsigned nbFiles) {
...
-        }
-        strcpy(filename, srcFilename);
-    } else {
-        filename = (char*) malloc((strlen(filenameBegin+1)+1) * sizeof(char));
-        if (!filename) {
-            EXM_THROW(30, "zstd: %s", strerror(errno));
-        }
-        strcpy(filename, filenameBegin+1);
-    }
+    filenameStart = extractFilename(path, separator);
```
Comment (terrelln) [inline programs/fileio.c]: Unrelated to this patch, since it doesn't change behavior, but what is the desired behavior if the source directory is `path/to/my/file/`?

I would expect the output paths to be relative to `path/to/my/file/`, but I believe the current behavior is to output `outdir/to/my/file`.

Comment (Cyan4973) [reply programs/fileio.c]: The functionality developed by @senhuang42 is called `--output-dir-flat`.

It places every compressed file, whatever its source directory, into a single, flat, target directory. It has the advantage of being simpler, both to describe and implement. 

It has the disadvantage of only allowing one file with a given name, so multiple files with same name from different directories will overwrite each other (there will be a prompt, similar to what happens today when overwriting a file).

Another command was considered, `--output-dir-mirror`, which would reproduce the source directory tree into a target directory. This patch was dropped for the time being. It could be resumed after `v1.4.4` release.
